### ISSUE TITLE
Converter registration in Session fails for repeated registration of any converter

### ISSUE THREAD

**Issue by xenji** [pomm-project/Foundation#38]
Setup:
- PHP 5.6
- Thrift (via overblog/thrift bundle)
- The thrift server is started as socket server in an event loop

Method (simplified for the use-case):
```php
    /**
     * @param string $name
     *
     * @return bool
     * @author Mario Mueller
     */
    public function existsByName($name)
    {
        $session = $this->db->getSession($this->userSession);
        $userModel = new UsersModel();
        $userModel->initialize($session); // This fails for repeated calls in the same event loop
        return $userModel->existWhere('username=$*', [$name]);
    }
```

Problem: 
Due to the event loop, the session does not get unloaded, so any custom converters are able to be added again. The `initialize(Session $session)` function mutates the session-related converters, this leads to an exception on the second request to the thrift server. 

Desired behaviour:
The session should check for the presence of the converter to add and either skip it or add it and make the list unique afterwards (what ever is more performant).

**Comment by xenji** [comment]
Maybe the problem is more narrow than described above. 

This code is taken from: `\PommProject\ModelManager\Model\Model::initialize` and the interesting part is the call to `\PommProject\Foundation\Converter\ConverterHolder::registerConverter`. There is no way to pass the strict parameter of `registerConverter` when calling `initialize`. 

```php
$session->getPoolerForType('converter')
            ->getConverterHolder()
            ->registerConverter(
                $this->flexible_entity_class,
                new PgEntity(
                    $this->flexible_entity_class,
                    $this->getStructure()
                ),
                [
                    $this->getStructure()->getRelation(),
                    $this->flexible_entity_class,
                ]
// HERE should be something like $strict, which then must be taken from the signature of ::initialize.
        );
```

A possible solution could be (and then it is more an issue of the model manager, not of foundation) to extend `public function initialize(Session $session)` to `public function initialize(Session $session, $strict = true)` in `\PommProject\ModelManager\Model\Model`.

**Comment by xenji** [comment]
I've added a pull request for the model manager that might solve the problem. This is subject to discussion. Looking forward to any comments ;)

**Comment by chanmix51** [comment]
Hello Mario,

Sorry for the lag. It am not sure I get the point of what you are trying to do. If the code in the first comment is a controller, why don't you use the getModel method ?

```php
<?php // controller.php
//…
   /**
     * @param string $name
     *
     * @return bool
     * @author Mario Mueller
     */
    public function existsByName($name)
    {
        return $this->db
            ->getSession($this->userSession)
            ->getModel('\Namespace\To\UsersModel')
            ->findWhere('username = $*', [$name]);
    }
```
The session pooler is made to avoid having to manage client registration to the session.

I am missing something ?

**Comment by xenji** [comment]
Ok, seems to me my bad in that case ^^. 

Mighty it be a good idea to control the strictness for the model init function anyway?

**Comment by xenji** [comment]
So the code could be...
```php
$session = $this->db->getSession($this->userSession);
$userModel = $session->getModel(UsersModel::class);
return $userModel->existWhere('username=$*', [$name]);
```

**Comment by xenji** [comment]
I've closed the pull request. It is unnecessary due to the way you provided here.

**Comment by chanmix51** [comment]
In fact, the POOLer managERS (poolers) have the responsability to handle these cases for us. By default, the pool is empty so the poolers create clients and register them to the session. If a client is needed several times, it is then fetched from the pool, there is no need to control a strictness unless the pooler's business rules are different (which might be the case) but they are handled by the pooler itself, it is its responsability to do so.

Thank you very much for the static attribute I was not aware of!

**Comment by xenji** [comment]
Ok, understood. Thanks for the detailed explanation! You mean the `::class`? Please keep in mind, that it is available since 5.5. See http://php.net/manual/de/migration55.new-features.php#migration55.new-features.class-name

**Comment by chanmix51** [comment]
I was looking for the minimum PHP version support, thank you :+1:
